Expand buildpack author concepts pages (#739)
buildpacks/docs: content/docs/for-buildpack-authors/concepts/layer.md

weight=99
+++

> `Dependency layers` are semantically meaningful layers that are contributed by one or more buildpacks during the building phase, one for each dependency.

Reviewer: Should this say `Dependency layers` or, more simply, `Layers`?

Author: Don't we have different layers? if not, then why the title reads as "dependency layers"?

Reviewer 2: We might first introduce the concept of an OCI image layer - see [here](https://github.com/opencontainers/image-spec/blob/main/layer.md) for a technical description which we might water down a little. It could be thought of as a filesystem changeset and some accompanying metadata. The ordering of layers usually matters for OCI images more generally but for buildpack contributed layers it does not (this is part of the beauty of our spec).

Then we might talk about the different kinds of layers you can have when building with buildpacks.
* There are base image layers (build and run), which are the layers that make up the underlying OS for the build-time container and the runtime container, respectively.
* Then you have buildpack-contributed dependency layers, which contain application dependencies (think: your language runtime & libraries).
* Then you have the application layer(s), which are usually some sort of transformation of the original source code that was provided (such as a compiled executable). We don't talk about "slices" here but we could link to it, it is a way to divide your application into many layers in order to make updates faster.
* Finally you have CNB-contributed layers, such as the SBOM, the launcher executable, and some configuration for the launcher.

Seeing as how this is buildpack author focused, we don't need to dwell on the first and last type of layer. But we can mention them to provide context.

The main point for buildpack authors is that by keeping layers small they can make them re-usable and composable. So I could have one layer that provides my language runtime and another that provides my package manager. The classic example is node (language) with either yarn or npm. This reduces duplication in what buildpack authors need to maintain.

Author: added

> `Dependency layers` are semantically meaningful layers that are contributed by one or more buildpacks during the building phase, one for each dependency.

Reviewer 2: We should link to https://buildpacks.io/docs/for-buildpack-authors/how-to/write-buildpacks/create-layer/ somewhere on this page (and vice versa)

Author: linked at the bottom as a note

> `Dependency layers` are semantically meaningful layers that are contributed by one or more buildpacks during the building phase, one for each dependency.

Author: updated

<!--more-->

![builder](/images/builder.svg)

> As seen in the image above, buildpacks read application source code and build dependency layers. Multiple buildpacks can contribute a subset of an app's required dependencies that are added as subdirectories under the `CNB_LAYERS_DIR` directory. These dependencies then get exported as layers in the final app image or build cache.

Author: updated

> The `build` phase runs the build binary of each buildpack that outputs zero or more layers into `$(CNB_LAYERS_DIR)/<buildpack-id>` and writes metadata for each layer as TOML files in that directory. During the `export` phase, all layers created by the buildpacks are either cached or added to the app final image.

Reviewer: ```suggestion
The `build` phase runs the build binary of each buildpack that outputs zero or more layers into `$(CNB_LAYERS_DIR)/<buildpack-id>` and writes metadata for each layer as TOML files in that directory. During the `export` phase, all layers created by the buildpacks are either cached or added to the output application image.
```

Author: updated

> The `build` phase runs the build binary of each buildpack that outputs zero or more layers into `$(CNB_LAYERS_DIR)/<buildpack-id>` and writes metadata for each layer as TOML files in that directory. During the `export` phase, all layers created by the buildpacks are either cached or added to the output application image.

Reviewer 2: Small nit, sometimes we say TOML and sometimes we say `toml`. No preference as to which we choose but we should be consistent where possible.

Author: updated all to toml

weight=99
+++

`Dependency layers` are semantically meaningful layers that are contributed by one or more buildpacks during the `build` phase, one for each dependency.

<!--more-->

![builder](/images/builder.svg)

> As seen in the image above, buildpacks read application source code and create dependency layers. Each buildpack can contribute a subset of an app's required dependencies; added as subdirectories under the `CNB_LAYERS_DIR` directory. These dependencies are then exported as layers in the final app image or build cache.

Reviewer 2: ```suggestion
As seen in the image above, buildpacks read application source code and create dependency layers. Each buildpack can contribute a subset of an app's required dependencies, added as subdirectories under the `CNB_LAYERS_DIR` directory. These dependencies are then exported as layers in the final app image or build cache.
```

> The `build` phase runs the build binary of each buildpack that outputs zero or more layers into `$(CNB_LAYERS_DIR)/<buildpack-id>` and writes metadata for each layer as `toml` files in that directory. During the `export` phase, all layers created by the buildpacks are either cached or added to the output application image.

Reviewer 2: ```suggestion
The `build` phase runs the `/bin/build` binary of each buildpack, which outputs zero or more layers into `$(CNB_LAYERS_DIR)/<buildpack-id>` and writes metadata for each layer as `toml` files in that directory. During the `export` phase, all layers created by the buildpacks are either cached or added to the output application image.
```

Reviewer 2: Maybe an example of what the filesystem might look like after the build phase is over would be helpful? Such as:

```
layers/
├── some-buildpack-id
│   ├── some-layer
│   ├── some-layer.toml
│   ├── some-other-layer
│   └── some-other-layer.toml
└── some-other-buildpack-id
```

Author: added
